### PR TITLE
Add Leo context menu metrics

### PR DESCRIPTION
**Pull request by DJAndries**
Resolves https://github.com/brave/brave-browser/issues/36560

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

For `Brave.AIChat.ContextMenu.FreeUsages` and `Brave.AIChat.ContextMenu.PremiumUsages`:

0. Start with fresh profile
1. Select some text on a page, and use a Leo context menu option
2. Before opting in, ensure metric does not exist in local state
3. Opt in to Leo. Ensure the free usages metric exists with the correct value.
4. Use the context menu options multiple times, ensure the metric count updates
5. Activate premium. Use the context menu again. Ensure the premium metric exists with the correct value, and the free metric is gone.

For `Brave.AIChat.ContextMenu.MostUsedAction`:

0. Start with fresh profile
1. Select some text on a page, and use a Leo context menu option
2. Before opting in, ensure metric does not exist in local state
3. Opt in to Leo. Ensure the metric exists with the correct value.
4. Use various context menu options, and ensure the metric value updates accordingly.

For `Brave.AIChat.ContextMenu.LastUsageTime`:

0. Start with fresh profile
1. Select some text on a page, and use a Leo context menu option
2. Before opting in, ensure metric does not exist in local state
3. Opt in to Leo. Ensure the metric exists with the value `1`.
4. Close browser, advance system time by a couple weeks, open browser, ensure metric value is correct.
5. Use context menu again, ensure value updates back to `1`.

### CODE REVIEW

**Comment by iefremov** [comment]
it looks, but could be very fragile without a browser test

**Comment by DJAndries** [comment]
> it looks good, but could be very fragile without a browser test

@iefremov updated

File: components/time_period_storage/time_period_storage.h [hunk 1]
```diff
@@ -61,6 +66,7 @@ class TimePeriodStorage {
 
   const raw_ptr<PrefService> prefs_;
   const char* pref_name_ = nullptr;
+  const char* dict_key_ = nullptr;
```
[the remark this answers is not in the document]
Comment (fmarier) [reply components/time_period_storage/time_period_storage.h:69]: False positive: https://github.com/brave/security-action/issues/537

File: components/time_period_storage/time_period_storage_unittest.cc [hunk 2]
```diff
@@ -15,15 +15,19 @@
...
 class TimePeriodStorageTest : public ::testing::Test {
  public:
   TimePeriodStorageTest() {
-    pref_service_.registry()->RegisterListPref(kPrefName);
+    pref_service_.registry()->RegisterListPref(kListPrefName);
+    pref_service_.registry()->RegisterDictionaryPref(kDictPrefName);
   }
 
-  void InitStorage(size_t days) {
+  void InitStorage(size_t days, const char* dict_key) {
```
Comment (iefremov) [inline components/time_period_storage/time_period_storage_unittest.cc:30]: cannot you make the parameter optional with ` = nullptr` ?

File: components/time_period_storage/time_period_storage.h [hunk 3]
```diff
@@ -26,10 +26,15 @@ class TimePeriodStorage {
...
                     size_t period_days);
+  TimePeriodStorage(PrefService* prefs,
+                    const char* pref_name,
+                    const char* dict_key,
+                    size_t period_days);
 
   // For tests.
   TimePeriodStorage(PrefService* prefs,
                     const char* pref_name,
+                    const char* dict_key,
```
Comment (iefremov) [inline components/time_period_storage/time_period_storage.h:37]: without digging the code it's not really clear what is it for, can you add a comment?

**Review by iefremov — APPROVED** [review]

**Pull request merged**